pkg/controller/node/cidr_allocator.go

```diff
@@ -17,144 +17,169 @@ limitations under the License.
 
 	ra := &rangeAllocator{
-		clusterCIDR:     clusterCIDR,
-		clusterIP:       clusterCIDR.IP.To4(),
-		clusterMaskSize: clusterMaskSize,
-		subNetMaskSize:  subNetMaskSize,
-		maxCIDRs:        1 << uint32(subNetMaskSize-clusterMaskSize),
-		nextCandidate:   0,
+		cidrs:                 newCIDRSet(clusterCIDR, subNetMaskSize),
+		client:                client,
+		clusterCIDR:           clusterCIDR,
+		nodeCIDRUpdateChannel: make(chan nodeAndCIDR, cidrUpdateQueueSize),
+		recorder:              recorder,
 	}
-	return ra
-}
 
-func (r *rangeAllocator) AllocateNext() (*net.IPNet, error) {
-	r.lock.Lock()
-	defer r.lock.Unlock()
-
-	nextUnused := -1
-	for i := 0; i < r.maxCIDRs; i++ {
-		candidate := (i + r.nextCandidate) % r.maxCIDRs
-		if r.used.Bit(candidate) == 0 {
-			nextUnused = candidate
-			break
-		}
+	if serviceCIDR != nil {
+		ra.filterOutServiceRange(serviceCIDR)
+	} else {
+		glog.V(0).Info("No Service CIDR provided. Skipping filtering out service addresses.")
 	}
-	if nextUnused == -1 {
-		return nil, errCIDRRangeNoCIDRsRemaining
+	for i := 0; i < cidrUpdateWorkers; i++ {
+		go func(stopChan <-chan struct{}) {
+			for {
+				select {
+				case workItem, ok := <-ra.nodeCIDRUpdateChannel:
+					if !ok {
+						glog.Warning("NodeCIDRUpdateChannel read returned false.")
+						return
+					}
+					ra.updateCIDRAllocation(workItem)
+				case <-stopChan:
+					glog.V(0).Info("StopChannel is closed.")
+					return
+				}
+			}
+		}(wait.NeverStop)
 	}
-	r.nextCandidate = (nextUnused + 1) % r.maxCIDRs
-
-	r.used.SetBit(&r.used, nextUnused, 1)
-
-	j := uint32(nextUnused) << uint32(32-r.subNetMaskSize)
-	ipInt := (binary.BigEndian.Uint32(r.clusterIP)) | j
-	ip := make([]byte, 4)
-	binary.BigEndian.PutUint32(ip, ipInt)
 
-	return &net.IPNet{
-		IP:   ip,
-		Mask: net.CIDRMask(r.subNetMaskSize, 32),
-	}, nil
+	return ra
 }
 
-func (r *rangeAllocator) Release(cidr *net.IPNet) error {
-	used, err := r.getIndexForCIDR(cidr)
+// AllocateOrOccupyCIDR looks at the given node, assigns it a valid CIDR
+// if it doesn't currently have one or mark the CIDR as used if the node already have one.
+func (r *rangeAllocator) AllocateOrOccupyCIDR(obj interface{}) error {
+	node := obj.(*api.Node)
+	if node.Spec.PodCIDR != "" {
+		_, podCIDR, err := net.ParseCIDR(node.Spec.PodCIDR)
+		if err != nil {
+			return fmt.Errorf("failed to parse node %s, CIDR %s", node.Name, node.Spec.PodCIDR)
+		}
+		if err := r.cidrs.occupy(podCIDR); err != nil {
+			return fmt.Errorf("failed to mark cidr as occupied: %v", err)
+		}
+		return nil
+	}
+	podCIDR, err := r.cidrs.allocateNext()
 	if err != nil {
-		return err
+		recordNodeStatusChange(r.recorder, node, "CIDRNotAvailable")
+		return fmt.Errorf("failed to allocate cidr: %v", err)
 	}
 
-	r.lock.Lock()
-	defer r.lock.Unlock()
-	r.used.SetBit(&r.used, used, 0)
-
+	glog.V(10).Infof("Putting node %s with CIDR %s into the work queue", node.Name, podCIDR)
+	r.nodeCIDRUpdateChannel <- nodeAndCIDR{
```

Comment:
IIRC, this channel was added because of performance reasons (to avoid blocking processing next nodes, on updates sent to apiserver, right? Can you please add a comment why do we have it (probably at the struct level).